proposals/infra/INF-0004-inproc-outofproc-compiler-library-support.md

```diff
@@ -0,0 +1,113 @@
+<!-- {% raw %} -->
+
+# Support for in-process, Out-of-process, or both for the c style compiler api
+
+* Proposal: [NNNN](INF-NNNN-inproc-outofproc-compiler-api-support.md)
+* Author(s): [Cooper Partin](https://github.com/coopp)
+* Sponsor: [Cooper Partin](https://github.com/coopp)
+* Status: **Under Consideration**
+* Impacted Project(s): (Clang)
+
+* Issues:
+
+## Introduction
+
+An effort is underway to bring compilation support into clang for HLSL based
+shaders.  A C style api will be built to enable applications to compile
+a shader from their own processes in addition to being able to launching the
+clang process.
+
+This document is to propose building only the out-of-process support which
+gives the caller a more performant and secure solution.
+
+Visual Studio uses an out of process architecture for compilation which helps
+better utilize the power of the machine performing the compilation as well as
+isolating the code editor from any crashes that may occur during compilation.
+
+### Overview of api hosting differences
+Here are two common ways a library can be loaded and used.
+
+* **In Process** - An application calls an api that loads compilation support
+into the application's process. All work is performed within the application's
+process.
+
+* **Out of Process** - An application calls an api that launches one or more
+processes to perform the compilation. All work is performed outside the
+calling application's process.
+
+## Motivation
+
+### Why prefer an out of process solution?
+
+* Process creation on Windows based systems is expensive.  Launching clang.exe
+over and over again to compile shaders becomes a lengthy operation on 
+Windows platforms.
+
+* Security is a concern when compiling a shader in a process space that is
+considered protected.  An out of process design would move that concern to an
+isolated process.
+
+* In-process support using multiple threads have known thread-safety concerns
+for the clang compiler.  The clang compiler holds some state at compile time
+and some of that state could leak to other compilation sessions if it is not
+protected. This issue was also called out when an experiment called
+[llvm-buildozer](https://reviews.llvm.org/D86351) was created.  The author was
+interested in multiple threads re-entering the main compiler entry point in an
+effort to reduce compilation times. The effort showed that the unsafe bits
+could be fixed up and made thread safe allowing the experiment to be built.
+I have not checked to see the thread safe fixed bits are in the current llvm
+source.
+
+## Proposed solution
```

Comment:
I think many of the details in here would belong in a detailed design section. I'm going to try and avoid reviewing the details here based on the experience with https://github.com/microsoft/hlsl-specs/pull/230.